src/ucp/dt/dt_iov.c

```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2015.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+#include "dt_iov.h"
+
+#include <ucs/debug/log.h>
+
+#include <string.h>
+
+
+void ucp_dt_iov_memcpy(void *dest, const ucp_dt_iov_t *iov, size_t length,
+                       size_t *iov_offset, size_t *iovcnt_offset)
+{
+    size_t item_len, length_shift, item_len_to_copy;
+    size_t length_it = 0;
+
+    ucs_assert(length > 0);
+    while (length_it < length) {
+        length_shift = 0;
+        item_len = iov[*iovcnt_offset].length;
+
+        /* make item upper copy limit */
+        if ((length_it + (item_len - *iov_offset)) > length) {
+            length_shift = (length_it + (item_len - *iov_offset)) - length;
```

Comment:
@yosefe these are always bigger than 0 because they are size_t types. Line 25 guarantee the `length_shift` will be "positive if ssiz_t".
I agree - these is a bit complicated algorithm :-) I tried to avoid several loops and do everything in one place.